### PR TITLE
[RLlib] Issue 30139: PolicyServerInput OOMs if incoming samples cannot be handled in time (e.g. when too many clients are connected).

### LINKED ISSUE #30139: [RLlib] PolicyServerInput memory leak
### What happened + What you expected to happen

`PolicyServerInput `keeps on storing samples in its `samples_queue`, no matter what the rate of sample generation is. If samples come in quicker than they are consumed, this leads to an ever-increasing memory usage. No warnings or anything were reported in the process of increasing memory. I had to debug why memory was ever-increasing and debugging the culprit took me several hours to track down.

At the very least I would expect some warning logging that the `samples_queue` is reaching very high memory usage (500GB), in my case in under 24 hours.

This is a well-known producer-consumer problem, and one solution would be to implement some kind of back-pressure mechanism. Alternatively, the queue size could be reduced in size when a certain 'max-size' parameter is reached when adding new samples. As a work-around I had to reduce the number of clients generating samples.

### Versions / Dependencies

Ray: 2.0.0
Python: 3.8.10
Ubuntu: 20.04.05

### Reproduction script

I would expect you can reproduce this if you start from the [Cartpole Server example](https://github.com/ray-project/ray/blob/master/rllib/examples/serving/cartpole_server.py). Create a big network, while setting `num_sgd_iter` to a high number. Create multiple clients. Track the `self.samples_queue` size in the `PolicyServerInput`. If it starts to increase consistently, you will run out of memory.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by sven1977**
Signed-off-by: sven1977 <svenmika1977@gmail.com>

Issue 30139: PolicyServerInput OOMs if incoming samples cannot be handled in time (e.g. when too many clients are connected).

Closes #30139 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/policy_server_input.py [hunk 1]
```diff
@@ -64,27 +66,42 @@ class PolicyServerInput(ThreadingMixIn, HTTPServer, InputReader):
...
+            max_queue_size: The maximum size for the sample queue. Once full, will
+                purge (throw away) 50% of all samples, oldest first, and continue.
         """
 
         self.rollout_worker = ioctx.worker
-        self.samples_queue = queue.Queue()
+        # Protect ourselves from having a bottleneck on the server (learning) side.
+        # Once the queue (deque) is full, we throw away 50% (oldest
+        # samples first) of the samples, warn, and continue.
+        self.samples_queue = deque(maxlen=max_sample_queue_size)
```
Comment (avnishn) [inline rllib/env/policy_server_input.py:104]: ah its good that we're bounding this now.

File: rllib/env/policy_server_input.py [hunk 2]
```diff
@@ -176,16 +193,22 @@ def get_metrics(self) -> List[RolloutMetrics]:
 
     @override(InputReader)
     def next(self):
-        return self.samples_queue.get()
+        # Blocking wait until there is something in the deque.
+        while len(self.samples_queue) == 0:
+            time.sleep(0.1)
```
Comment (avnishn) [inline rllib/env/policy_server_input.py:198]: its important to know that time.sleep is equivalent to yield in terms of letting go of control of the processor to other threads...

File: rllib/env/policy_server_input.py [hunk 3]
```diff
@@ -64,27 +66,42 @@ class PolicyServerInput(ThreadingMixIn, HTTPServer, InputReader):
...
-        server using rllib.env.PolicyClient.
+        server using rllib.env.PolicyClient. You can increase the number of available
+        connections (ports) by setting num_rollout_workers to a larger number. The ports
+        used will then be `port` + the worker's index.
 
         Args:
             ioctx: IOContext provided by RLlib.
             address: Server addr (e.g., "localhost").
             port: Server port (e.g., 9900).
+            max_queue_size: The maximum size for the sample queue. Once full, will
```
Comment (avnishn) [inline rllib/env/policy_server_input.py:96]: can this be infinity?

File: rllib/env/policy_server_input.py [hunk 4]
```diff
@@ -217,7 +240,14 @@ def report_data(data):
 
         batch = data["samples"]
         batch.decompress_if_needed()
-        samples_queue.put(batch)
+        samples_queue.append(batch)
+        # Deque is full -> purge 50% (oldest samples)
+        if len(samples_queue) == samples_queue.maxlen:
```
Comment (avnishn) [inline rllib/env/policy_server_input.py:245]: hmm what happens here if someone passes infinity or 0 or none for queue maxlen

Comment (avnishn) [reply rllib/env/policy_server_input.py:245]: will this cause an error?

**Review by avnishn — COMMENTED** [review]
I have some concerns about whether making the queue max length infinity or 0 will cause errors when checking to see if the queue should be purged.

Could you p pls address them? Otherwise, LGTM

**Comment by MattiasDC** [comment]
I would like this PR to be in for 2.3.0, is it possible to give this one a review? :)

**Comment by MattiasDC** [comment]
Removing stale label..

**Comment by MattiasDC** [comment]
Removing stale label @avnishn , could you have a look?

**Comment by DenysAshikhin** [comment]
Just wanted to add my own experience that implementing this in my own project fixed a memory leak I had with policy server/client - so it would be great if this could be pulled in!

**Comment by kouroshHakha** [comment]
@sven1977 Can you check the PR to see if it can still get merged in?

**Comment by DenysAshikhin** [comment]
@kouroshHakha @sven1977 any update for this? It would be greatly appreciated by me! (and probably ninja fix any issues others many not even realise using policy_server/client)

**Review by kouroshHakha — APPROVED** [review]

**Comment by DenysAshikhin** [comment]
Awesome to see this pulled in! (I can finally update my ray once this hits nightle releases.) Speaking of which, how can I tell once this is in a nightly release?

**Comment by kouroshHakha** [comment]
@DenysAshikhin One way is to look for the commit to see if it's passed this commit number in the history.
```
import ray
print(ray.__commit__)
```

**Comment by DenysAshikhin** [comment]
Awesome, thanks!

**Pull request merged**